designs/simultaneous-edit/README.md

```diff
@@ -0,0 +1,121 @@
+---
+state: draft
+---
+
+# Simultaneous Flow Editing
+
+## Summary
+
+With current Node-RED, multiple clients can connect to the same runtime and can edit the same flow independently.  However, editing results by multiple clients are not synchronized, so conflicts of flows occur during deployment of the flow.  There is a feature to support merging flow conflicts referring node property values, but it is difficult for end users to merge node properties in proper way.
+
+![conflict-detection.png](/Users/nisiyama/src/Curry/Git/designs/designs/simultaneous-edit/images/conflict-detection.png)
+
+![review-change.png](/Users/nisiyama/src/Curry/Git/designs/designs/simultaneous-edit/images/review-change.png)
+
+Therefore, this design document proposes a new Node-RED editor feature that supports multiple clients to edit the same flow simultaneously.
+
+## Authors
+
+- Hiroyasu Nishiyama
+
+## Proposal
+
+Here, we consider the following two methods as candidates for simultaneous editing support of a Node-RED flow:
+
+1. Synchronize editing actions among clients,
+
+2. Allow locking a tab by a client to make single client can only edit the tab.
+
+### Method1: Edit Action Synchronization
+
+#### 1-a: Requirements
+
+- The flow edit result of a client is immediately reflected to the flow edited on workspaces of other clients.
+
+#### 1-b: Issues
+
+- How to reflect flow editing state among clients?
+
+- How to handle other kind of state changes of Node-RED including editor and runtime (e.g. node installation, execution result, etc.)?
+
+#### 1-c: Implementation
+
+- Extends the edit history information currently used for *undo* / *redo* of flow editing operations, and sends editing action information to other clients via the runtime.
```

Comment:
I think I have a branch of code locally that started looking at how to generate [JSON Patch](http://jsonpatch.com/) fragments from the history component. That seemed to be the most sensible way to approach this - as it should be possible to apply patches in a well defined order (and also has some well-known algorithms for resolving out-of-order merges).

Reply:
My assumption was to serialize the editing actions by communicating with runtime APIs.
Your suggestion using JSON patch may be useful if we deal with it as client-side merge process.  However, I feel it will be difficult to guarantee that multiple clients will get the same editing results.

Reply:
JSON Patch is one possible format - but the key is the history component would be the central place to know that a change has happened, and that change should be communicated to other editors. When the other editors receive the details of that event, they need to apply them to the local configuration - translating the event from whatever format it is in back into whatever modification to internal state in the editor is needed.

Reply:
I agree.  My assumption is that the runtime manages the edit histories to keep the order consistent for all clients.

Reply:
Shifting the edit history to the runtime is a major change to how the editor works. We'd need to see a lot more detail on how you think that would work. Resolving concurrent edit state between multiple parties is a very complex task - there are countless research papers on the topic. This is the part that I'm most interested in - we cannot underestimate the difficulty in getting it right.

For example, how do you resolve conflicting changes to the same resource. How do you manage ordering of changes. How do you manage undoing a change.

But this still remains an academic conversation untill we under stand how we handle the deploy button issue.

Reply:
Thanks for your comment. I have updated the design document.

Regarding the issue of the deploy button, I would like to suggest a method that all users to save the fix and then deploy it.
Also, for edit action conflicts, I would like to suggest serialized application of operations based on atomic application of each edit actions.